### ISSUE TITLE
Link errors using firebase_firestore in CMake on macOS desktop

### ISSUE THREAD

**Issue by mark-grimes** [firebase/firebase-cpp-sdk#53]
### Please fill in the following fields:
Pre-built SDK from the [website](https://firebase.google.com/download/cpp) or open-source from this repo: website
Firebase C++ SDK version: 6.12.0
Firebase plugins in use (Auth, Database, etc.): Firestore (not relevant but also using Auth and Storage)
Additional SDKs you are using (Facebook, AdMob, etc.): 
Platform you are using the C++ SDK on (Mac, Windows, or Linux): Mac
Platform you are targeting (iOS, Android, and/or desktop): desktop

### Please describe the issue here:
(Please list the full steps to reproduce the issue. Include device logs, Unity logs, and stack traces if available.)

Attempting to link to the `firebase_firestore` CMake target gives
```
Undefined symbols for architecture x86_64:
  "_SCNetworkReachabilityCreateWithAddress", referenced from:
      firebase::firestore::remote::ConnectivityMonitorApple::ConnectivityMonitorApple(std::__1::shared_ptr<firebase::firestore::util::AsyncQueue> const&) in libfirebase_firestore.a(d43fad9bb2ebb8a979aa11feb868dab3_connectivity_monitor_apple.o)
  "_SCNetworkReachabilityGetFlags", referenced from:
      firebase::firestore::remote::ConnectivityMonitorApple::ConnectivityMonitorApple(std::__1::shared_ptr<firebase::firestore::util::AsyncQueue> const&) in libfirebase_firestore.a(d43fad9bb2ebb8a979aa11feb868dab3_connectivity_monitor_apple.o)
  "_SCNetworkReachabilitySetCallback", referenced from:
      firebase::firestore::remote::ConnectivityMonitorApple::ConnectivityMonitorApple(std::__1::shared_ptr<firebase::firestore::util::AsyncQueue> const&) in libfirebase_firestore.a(d43fad9bb2ebb8a979aa11feb868dab3_connectivity_monitor_apple.o)
  "_SCNetworkReachabilitySetDispatchQueue", referenced from:
      firebase::firestore::remote::ConnectivityMonitorApple::ConnectivityMonitorApple(std::__1::shared_ptr<firebase::firestore::util::AsyncQueue> const&) in libfirebase_firestore.a(d43fad9bb2ebb8a979aa11feb868dab3_connectivity_monitor_apple.o)
      firebase::firestore::remote::ConnectivityMonitorApple::~ConnectivityMonitorApple() in libfirebase_firestore.a(d43fad9bb2ebb8a979aa11feb868dab3_connectivity_monitor_apple.o)
  "absl::str_format_internal::FormatPack(absl::str_format_internal::UntypedFormatSpecImpl, absl::Span<absl::str_format_internal::FormatArgImpl const>)", referenced from:
      f_b_grpc_chttp2_transport::f_b_grpc_chttp2_transport(grpc_channel_args const*, grpc_endpoint*, bool, grpc_resource_user*) in libfirebase_firestore.a(ebe3cf7b618c76d718a4bf7736e66541_chttp2_transport.o)
  "bool absl::str_format_internal::FormatArgImpl::Dispatch<char const*>(absl::str_format_internal::FormatArgImpl::Data, absl::str_format_internal::FormatConversionSpec, void*)", referenced from:
      f_b_grpc_chttp2_transport::f_b_grpc_chttp2_transport(grpc_channel_args const*, grpc_endpoint*, bool, grpc_resource_user*) in libfirebase_firestore.a(ebe3cf7b618c76d718a4bf7736e66541_chttp2_transport.o)
ld: symbol(s) not found for architecture x86_64
```
Linking to the SystemConfiguration framework gets rid of the first four, so it would be helpful to have that listed in the [documentation](https://firebase.google.com/docs/cpp/setup#os-x-libraries).

I can't fix the last two symbols though. I presume they're from the [abseil.io](abseil.io) library, so I built that.  The symbols seem to be in the library:
```
username$ nm -Ug ../../abseil_install/lib/*.a
<snip>
../../abseil_install/lib/libabsl_str_format_internal.a(bind.cc.o):
<snip>
0000000000000ef0 T __ZN4absl14lts_2020_02_2519str_format_internal10FormatPackENS1_21UntypedFormatSpecImplENS0_4SpanIKNS1_13FormatArgImplEEE

<snip>
../../abseil_install/lib/libabsl_str_format_internal.a(arg.cc.o):
0000000000002a20 T __ZN4absl14lts_2020_02_2519str_format_internal13FormatArgImpl8DispatchINS0_11string_viewEEEbNS2_4DataENS1_20FormatConversionSpecEPv
0000000000002070 T __ZN4absl14lts_2020_02_2519str_format_internal13FormatArgImpl8DispatchINS0_6int128EEEbNS2_4DataENS1_20FormatConversionSpecEPv
00000000000022c0 T __ZN4absl14lts_2020_02_2519str_format_internal13FormatArgImpl8DispatchINS0_7uint128EEEbNS2_4DataENS1_20FormatConversionSpecEPv
0000000000000000 T __ZN4absl14lts_2020_02_2519str_format_internal13FormatArgImpl8DispatchINS1_7VoidPtrEEEbNS2_4DataENS1_20FormatConversionSpecEPv
<snip>
```
...but linking the `absl_str_format_internal` library in doesn't solve anything.  I presume name mangling stops the symbols being located.  Abseil and my Firebase code are compiled with Apple clang version 11.0.3 (clang-1103.0.32.29).

### Please answer the following, if applicable:
Have you been able to reproduce this issue with just the [Firebase C++ quickstarts](https://github.com/firebase/quickstart-cpp) ?

Not applicable - issue seems to be with project setup, not code. I haven't used anything other than `Firestore::GetInstance(...)` yet.

What's the issue repro rate? (eg 100%, 1/5 etc)

100%

**Comment by morganchen12** [comment]
The missing absl symbols are likely a result of the wrong version of abseil being linked into your application. If you `nm` the resulting binary, you should see most of the abseil symbols are present.

What version of abseil are you linking in? You should see it in the output of your CocoaPods installation, or in the resulting `Podfile.lock`.

**Comment by mark-grimes** [comment]
Configuration is with CMake, I thought the CocoaPods requirement was for the mobile builds.  Auth and Storage work fine before adding Firestore.

The abseil version I compiled and linked to was the git tag `20200225.1`, but after looking in the CocoaPods file you linked I tried `20190808` but get the same result.

**Comment by morganchen12** [comment]
I've been informed that CocoaPods is not required for Firestore C++ on macOS. I'll do some digging and get back to you with a more complete answer.

**Comment by mark-grimes** [comment]
Thanks.  Note that I don't want to link to abseil, it was just an attempt to solve the earlier problem.  If I `nm -gU` on the firestore C++ library (firebase_cpp_sdk/libs/darwin/x86_64/libfirebase_firestore.a) there are plenty of other abseil symbols.

**Comment by wilhuff** [comment]
First, sorry for the trouble here. The missing Abseil library has been fixed internally and is waiting on the next release (you can see it in cl/306284550 if you're a Googler--I can't wait until this all open source).

Your intuition on a workaround is a good one, but Abseil distributions come in two flavors: long-term support releases and regular live at head versions. The former contains an inline namespace including the LTS version (e.g. `absl::lts_2020_02_25::str_format_internal`) while the latter does not include an inline namespace (`absl::str_format_internal`).

(Incidentally, you can see this more clearly if you pipe the `nm` output through `c++filt`, which shows that your build of Abseil defines LTS symbols:

```
bool absl::lts_2020_02_25::str_format_internal::FormatArgImpl::Dispatch<absl::lts_2020_02_25::str_format_internal::VoidPtr>(absl::lts_2020_02_25::str_format_internal::FormatArgImpl::Data, absl::lts_2020_02_25::str_format_internal::FormatConversionSpec, void*)
```

The missing symbols are non-LTS.

The best workaround at this point is therefore to link to a head build from around the time of Firebase C++ SDK's release (Mar 5), which would be https://github.com/abseil/abseil-cpp/tree/cf3a1998e9d41709d4141e2f13375993cba1130e. The `StrFormat` API has been pretty stable so this is unlikely to cause problems. Unfortunately, I can't tell you exactly which version we used because the internal build and github don't align (again--I can't wait until this all open source).

**Comment by mark-grimes** [comment]
Thanks for the speedy response.  If I use the current abseil HEAD (1a02b7a2054c24f900dab796edb812f9260b51a6) I now get:
```
Undefined symbols for architecture x86_64:
  "absl::operator/(absl::uint128, absl::uint128)", referenced from:
      void absl::str_format_internal::(anonymous namespace)::ConvertedIntInfo::UnsignedToStringRight<absl::uint128>(absl::uint128, absl::str_format_internal::FormatConversionChar) in libabsl_str_format_internal.a(arg.cc.o)
  "absl::operator%(absl::uint128, absl::uint128)", referenced from:
      void absl::str_format_internal::(anonymous namespace)::ConvertedIntInfo::UnsignedToStringRight<absl::uint128>(absl::uint128, absl::str_format_internal::FormatConversionChar) in libabsl_str_format_internal.a(arg.cc.o)
ld: symbol(s) not found for architecture x86_64
```
So the original problem symbols are now found. Either I need to figure out a working commit from ~Mar 5 as you suggest, or add some other abseil libraries.  It'll be a few days before I get time to sort it out though, I'll post back then.

**Comment by wilhuff** [comment]
Those look like you're missing dependencies of str_format_internal, notably the int128 library. You can see the list in [the Abseil build](https://github.com/abseil/abseil-cpp/blob/cb52b05ea20e8d196f17b45a420c2d323e401324/absl/strings/CMakeLists.txt#L395).

You may find it easier to implement this workaround by pulling in the Abseil build directly into your CMake project--then the transitive closure of dependencies will be pulled into your link line automatically.

**Comment by wilhuff** [comment]
I'm glad to hear that you have a workaround and again I'm sorry that you had to do it. We will be publishing a next release that fixes this. Once we do we'll close this issue.

**Comment by mark-grimes** [comment]
Actually, just after posting that I was trying to fix the other issue (continuations not being called) and I stumbled across a newer version of Firebase - 6.14.0.  With that version it all links fine without that CMake snippet above.

One minor point is that it's difficult to find out what the latest version is.  "Releases" on the GitHub project is three minor versions out of date (6.11.0), and the only other way is to click on the link at the top of [Add Firebase to your C++ project](https://firebase.google.com/docs/cpp/setup)...  I have to start downloading the file to see the version in the filename - and then cancel the download if it's one I already have.

Closing since it works with 6.14.0.
